src/internal.c

```diff
@@ -5114,11 +5143,39 @@ int EccMakeKey(WOLFSSL* ssl, ecc_key* key, ecc_key* peer)
     #endif
     }
 
-    /* Handle async pending response */
+#ifdef WOLFSSL_ECDHE_NONBLOCK
+    if (ret == FP_WOULDBLOCK) {
+        ret = WC_PENDING_E;
+    }
+#endif /* WOLFSSL_ECDHE_NONBLOCK */
+
 #ifdef WOLFSSL_ASYNC_CRYPT
-    if (ret == WC_PENDING_E) {
-        ret = wolfSSL_AsyncPush(ssl, &key->asyncDev);
+    /* Handle async pending response */
+
+    /*
+     * Don't add a new async event if there's already a non-blocking operation
+     * in progress.
+     */
+    if (!nonblockInProg) {
+        if (ret == WC_PENDING_E) {
+            ret = wolfSSL_AsyncPush(ssl, &key->asyncDev);
+        }
+    }
+    else {
+    #ifdef WOLFSSL_ECDHE_NONBLOCK
+        if (ret == 0) {
+            /* Non-blocking operation complete. */
+            ret = wolfEventQueue_Remove(&ssl->ctx->event_queue, &ssl->asyncDev->event);
```

Comment:
It seems like the `wolfEventQueue_Remove` could be handled in the async code via the call to `wolfSSL_AsyncPop`?